### PR TITLE
Panorama Lookup commands

### PR DESCRIPTION
**Pull request by adambaumeister**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Description
Adds two new commands:
**panorama-lookup-route**
Searches the routing table on a firewall for the given IP address and returns the route details.
**panorama-lookup-zone**
Compares both the interface and routing table to make a determination of zone in which an IP address resides

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [X] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation

### CODE REVIEW

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 1]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+def panorama_route_lookup(dest_ip: str, virtual_router=None):
+    """
+    Given the provided ip address, looks up the outgoing interface and zone on the firewall.
+    """
+    if not VSYS:
+        raise Exception("The 'panorama-route-lookup' command is only relevant for a Firewall instance.")
+
+    response = panorama_get_routes(virtual_router)
+    if 'entry' not in response['response']['result']:
+        raise Exception("No routes returned from the Firewall.")
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4902]:
```suggestion
        raise DemistoException("No routes returned from the Firewall.")
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 2]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+        'GET',
+        params=params)
+
+
+def panorama_route_lookup(dest_ip: str, virtual_router=None):
+    """
+    Given the provided ip address, looks up the outgoing interface and zone on the firewall.
+    """
+    if not VSYS:
+        raise Exception("The 'panorama-route-lookup' command is only relevant for a Firewall instance.")
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4898]:
```suggestion
        raise DemistoException("The 'panorama-route-lookup' command is only relevant for a Firewall instance.")
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 3]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+
+
+def panorama_route_lookup_command():
+    """
+    Gets the outgoing interface from the Palo Alto Firewall route table
+    Must be run against a firewall instance
+    """
+    dest_ip = demisto.args().get("dest_ip")
+    vr = demisto.args().get("virtual_router", None)
+    demisto.results(panorama_route_lookup(dest_ip, vr))
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4937]: please use return_results method instead,
see this [doc](https://xsoar.pan.dev/docs/integrations/code-conventions#return_results)

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 4]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+            'Type': entryTypes['note'],
+            'ContentsFormat': formats['json'],
+            'Contents': r,
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': f'The IP {dest_ip} is in zone {r["zone"]}',
+            'EntryContext': {"Panorama.ZoneLookup(val.Name == obj.Name)": r}  # add key -> deleted: true
+        })
+        return r
+    else:
+        demisto.results(f"Could not map {dest_ip} to zone.")
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4973]: please use return_results method instead,
see this [doc](https://xsoar.pan.dev/docs/integrations/code-conventions#return_results)

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 5]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+def panorama_zone_lookup_command():
+    """
+    Gets the outgoing interface from the Palo Alto Firewall route table, and the list of interfaces
+    comparing the two
+    """
+    dest_ip = demisto.args().get("dest_ip")
+    vr = demisto.args().get("virtual_router", None)
+    route = panorama_route_lookup(dest_ip, vr)
+    if not route:
+        demisto.results(f"Could find a matching route to {dest_ip}.")
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4949]: please use return_results method instead,
see this [doc](https://xsoar.pan.dev/docs/integrations/code-conventions#return_results)

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 6]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+
+    r = {}
+    if "ifnet" in interfaces["response"]["result"]:
+        for entry in interfaces["response"]["result"]["ifnet"]["entry"]:
+            if entry["name"] == interface:
+                if "zone" in entry:
+                    r = {**entry, **route}
+
+    if r:
+        demisto.results({
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4963]: please use return_results method instead,
see this [doc](https://xsoar.pan.dev/docs/integrations/code-conventions#return_results)

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 7]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+
+def panorama_route_lookup(dest_ip: str, virtual_router=None):
+    """
+    Given the provided ip address, looks up the outgoing interface and zone on the firewall.
+    """
+    if not VSYS:
+        raise Exception("The 'panorama-route-lookup' command is only relevant for a Firewall instance.")
+
+    response = panorama_get_routes(virtual_router)
+    if 'entry' not in response['response']['result']:
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4901]:
```suggestion
    if 'entry' not in response['response'].get('result'):
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 8]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+        raise Exception("No routes returned from the Firewall.")
+    else:
+        routes = response['response']['result']['entry']
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4904]:
```suggestion
        if response.get('response') and response['response'].get('result') and response['response']['result'].get('entry'):
                routes = response['response']['result']['entry']
         else:
                 routes = []
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 9]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+
+    ip_addr = ipaddress.ip_address(dest_ip)
+    current_match = None
+    matched_route = None
+    for route in routes:
+        subnet_raw = route['destination']
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4910]:
```suggestion
        subnet_raw = route.get('destination')
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 10]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+                matched_route = route
+            # If this is a greater subnet
+            elif subnet.prefixlen > current_match.prefixlen:
+                current_match = subnet
+                matched_route = route
+
+    if matched_route:
+        return matched_route
+    else:
+        raise Exception("Route not found.")
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4927]:
```suggestion
        raise DemistoException("Route not found.")
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 11]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+    route = panorama_route_lookup(dest_ip, vr)
+    if not route:
+        demisto.results(f"Could find a matching route to {dest_ip}.")
+        return
+
+    interface = route["interface"]
+    interfaces = panorama_get_interfaces()
+
+    r = {}
+    if "ifnet" in interfaces["response"]["result"]:
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4956]:
```suggestion
    if "ifnet" in interfaces["response"].get("result"):
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 12]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+    comparing the two
+    """
+    dest_ip = demisto.args().get("dest_ip")
+    vr = demisto.args().get("virtual_router", None)
+    route = panorama_route_lookup(dest_ip, vr)
+    if not route:
+        demisto.results(f"Could find a matching route to {dest_ip}.")
+        return
+
+    interface = route["interface"]
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4952]:
```suggestion
    interface = route.get("interface")
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 13]
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+    interfaces = panorama_get_interfaces()
+
+    r = {}
+    if "ifnet" in interfaces["response"]["result"]:
+        for entry in interfaces["response"]["result"]["ifnet"]["entry"]:
+            if entry["name"] == interface:
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4958]:
```suggestion
            if entry.get("name") == interface:
```

**Review by adi88d — CHANGES_REQUESTED** [review]
Thanks for your contribution @adambaumeister!  it Looks great.
few notes for both `panorama-lookup-route` and `panorama-lookup-zone` commands:
- please return a human readable table containing all the output of the command + the raw response from the api, use `return_results` for this see the docs [here](https://xsoar.pan.dev/docs/integrations/code-conventions#return_results).
- Update the outputs of the commands, you can use `json-to-outputs` command in demisto-sdk for this.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py (hunk 5 above)
Comment (adambaumeister) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:4954]: Fixed

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py (hunk 3 above)
Comment (adambaumeister) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: Fixed

**Comment by adambaumeister** [comment]
I updated a couple of things here
* As requested, changed all teh demisto.results() calls to return_results instead (Way easier anyway!!!)
* Changed the route lookup logic to be based on the _test routing fib-lookup_ command instead - I realized that in a firewall with a lot of different paths the previous command may have returned misleading results
* Added outputs as requested and also added descriptions.
* Added some more, better tests

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 14]
```diff
@@ -4846,6 +4848,138 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
     )
 
 
+def panorama_get_routes(virtual_router=None):
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4851]: remove this method if you dont use it

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 15]
```diff
@@ -4846,6 +4848,138 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+    if "ifnet" in interfaces["response"].get("result"):
+        for entry in interfaces["response"]["result"]["ifnet"]["entry"]:
+            if entry.get("name") == interface:
+                if "zone" in entry:
+                    r = {**entry, **route}
+
+    if r:
+        # add the given dest_ip as another item int the object for simplicity
+        r["dest_ip"] = dest_ip
+        markdown = tableToMarkdown('Zone Lookup Results', r, headers=["dest_ip", "zone", "interface"])
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4969]: please add the rest of the headers name, ip, id, etc.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 16]
```diff
@@ -4846,6 +4848,138 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
...
+    """
+    if not VSYS:
+        raise DemistoException("The 'panorama-route-lookup' command is only relevant for a Firewall instance.")
+
+    dest_ip = demisto.args().get("dest_ip")
+    vr = demisto.args().get("virtual_router", "default")
+    r = panorama_route_lookup(dest_ip, vr)
+
+    r["dest_ip"] = dest_ip
+    markdown = tableToMarkdown('Route Lookup Results', r, headers=["dest_ip", "interface"])
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:4930]: please add the rest of the headers ip, src, metric, etc.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 17]
```diff
@@ -4370,7 +4372,110 @@ script:
...
+      description: 'Outgoing Interface Name'
+      type: String
+    - contextPath: Panorama.ZoneLookup.zone
+      description: 'Outgoing Zone Name'
+      type: String
+    - contextPath: Panorama.ZoneLookup.fwd
+      description: 'Forwarding Virtual Router'
+      type: String
+    - contextPath: Panorama.ZoneLookup.dyn-addr
+      description: ''
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:4443]: add description

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 18]
```diff
@@ -4370,7 +4372,110 @@ script:
...
+      description: 'Outgoing interface IPv6 addresses'
+      type: String
+    - contextPath: Panorama.ZoneLookup.tag
+      description: 'Outgoing interface VLAN tag'
+      type: String
+    - contextPath: Panorama.ZoneLookup.ip
+      description: 'Outgoing interface IP'
+      type: String
+    - contextPath: Panorama.ZoneLookup.id
+      description: ''
```
Comment (adi88d) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:4455]: add description

**Review by adi88d — CHANGES_REQUESTED** [review]
Nice! see my comments in the PR, please update the README file - use `generate-docs` in the SDK and add example for the commands  `panorama-lookup-route` and `panorama-lookup-zone`.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py (hunk 14 above)
Comment (adambaumeister) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: done!

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py (hunk 15 above)
Comment (adambaumeister) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: Resolved by removing the header spec altogether and letting tableToMarkdown figure it out (there are a number of headers that may or may not appear depending on the FW config, so I figured this was the most suitable way)

**Review by adi88d — CHANGES_REQUESTED** [review]
- please update the README file - use `generate-docs` in the SDK and add example for the commands  `panorama-lookup-route` and `panorama-lookup-zone`.
- update the `palo_alto_firewall_test_pb` playbook

**Comment by adambaumeister** [comment]
Hi, 

I tried updating the docs using generate-docs but it doesn't seem to work right. It always generates an empty "command example" string and no human readable  output.  I ran with the following command:

```
demisto-sdk generate-docs -i Panorama.yml -c "panorama-lookup-zone,panorama-lookup-route" --insecure
```

```
### panorama-lookup-route
***
Lookup a given IP address in the Firewall routing table


#### Base Command

`panorama-lookup-route`
#### Input

| **Argument Name** | **Description** | **Required** |
| --- | --- | --- |
| dest_ip | Desintation IP to lookup | Required |
| virtual_router | virtual router to check against | Optional |


#### Context Output

| **Path** | **Type** | **Description** |
| --- | --- | --- |
| Panorama.RouteLookup.dest_ip | String | The IP address tested |
| Panorama.RouteLookup.dp | String | Firewall dataplane |
| Panorama.RouteLookup.interface | String | Outgoing interface |
| Panorama.RouteLookup.ip | String | The next hop IP address |
| Panorama.RouteLookup.metric | String | Route metric |
| Panorama.RouteLookup.nh | String | next hop type |
| Panorama.RouteLookup.src | String | Outgoing interface IP |


#### Command Example
``` ```

#### Human Readable Output


```

It seems like you have to specify `-e command_examples`, but this tries to run the generation for every single command in the integration and a lot of them don't work on my test instance.

Should I manually populate these values?

Cheers
Adam

**Comment by adambaumeister** [comment]
Added an example playbook, updated the README and merged most recent changes from Upstream.

**Comment by ShirleyDenkberg** [comment]
@adi88d  Doc review completed.

**Review by adi88d — APPROVED** [review]

**Pull request merged**